ASTInfoGenerator/ast_info_generator.py

```diff
@@ -0,0 +1,190 @@
+output_file = sys.argv[2]
+
+with open(file_to_read) as ast:
+    cur_spec = None  # Function or Struct or None
+    cur_top_decl = None  # FunctionDecl or RecordDecl (strings)
+    cur_mid_decl = None  # ParmVarDecl or FieldDecl (strings)
+    param_index = None
+    field_index = None
+
+    ast_lines = ast.readlines()
+    specifier_manager = SpecifierManager()
+    annotation_manager = AnnotationManager()
+    struct_var_manager = StructVarManager()
+    decl_parser = DeclParser()
+
+    for expr in ast_lines:
+        expr = expr.strip()
+
+        first_letter_index = expr.find(next(filter(str.isalpha, expr)))
+        start_decl_index = first_letter_index
+        decl = ""
+        cur_decl_char = expr[start_decl_index]
+
+        try:
+            while (cur_decl_char != " "):
+                decl += cur_decl_char
+                start_decl_index += 1
+                cur_decl_char = expr[start_decl_index]
+        except IndexError:
+            # might be a null statement (looks like "<<<NULL>>>" in AST)
+
+            # TODO: write explicit check for null statements; lack of
+            # documentation on the formatting of
+            # null statements makes writing explicit checks difficult
+
+            field_index = None
+            continue
+
+        type_parsed = decl_parser.raw_ast_to_decl_type(expr, specifier_manager)
+        found_type = type(type_parsed)
+
+        if found_type is FunctionDecl:
+            specifier_manager.add_function(
+                type_parsed.get_fn_name(), type_parsed.get_ret_type())
+
+        elif found_type is RecordDecl:
+            specifier_manager.add_struct(type_parsed.get_struct_name())
+
+        elif found_type is FieldDecl:
+            spec = specifier_manager.get_specifiers()[-1]
+            if type(spec) is Struct:
+                field = Field(type_parsed.get_field_name(),
+                              type_parsed.get_field_index())
+                spec.add_field(field)
+
+        elif found_type is ParmVarDecl:
+            spec = specifier_manager.get_specifiers()[-1]
+            if type(spec) is Function:
+                param = Parameter(type_parsed.get_parm_index(),
+                                  type_parsed.get_parm_type())
+                spec.add_parameter(param)
+
+        elif found_type is TypedefDecl:
+            specifier_manager.get_struct(type_parsed.get_original_type_name()).add_typedef(
+                type_parsed.get_alias_name())
+
+        elif found_type is StructVarDecl:
+            struct_var_manager.add_struct_var(
+                type_parsed.get_struct_var_name(), type_parsed.get_struct_type_name())
+
+        elif found_type is AnnotateAttr:
+            anno = Annotation(type_parsed.get_annotation_type(
+            ), type_parsed.get_annotation_target(), type_parsed.get_annotation_methods())
+            annotation_manager.add_annotation(anno)
+
+    if DEBUG:
+        for spec in specifier_manager.get_specifiers():
```

Comment:
This debug printing code shouldn't be separate from the code that writes the result just below - if they ever diverge, it'll be extremely inconvenient to debug. Instead, structure the code in the following way:
* write a helper function that produces a (large) string
* make this code call that helper and then print the string
* make the output-writing code below call the same helper function and then write the string to a file